File: brique-24-sdk-auth/sdk/android/MolamAuth.kt
/**
 * Molam Auth SDK - Android (Kotlin)
 * Universal authentication SDK for Molam ID
 *
 * Version: 1.0.0
 * Author: Molam
 */

package com.molam.auth

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.*

data class MolamAuthConfig(
    val baseUrl: String,
    val apiKey: String? = null,
    val autoRefresh: Boolean = true
)

data class LoginRequest(
    val identifier: String,
    val password: String,
    val deviceId: String? = null,
    val deviceType: String = "android"
)

data class AuthResponse(
    val accessToken: String,
    val refreshToken: String,
    val expiresIn: Int,
    val tokenType: String,
    val user: User
) {
    data class User(
        val id: String,
        val email: String?,
        val phoneE164: String?
    )
}

class MolamAuth private constructor(
    private val context: Context,
    private val config: MolamAuthConfig
) {
    private val prefs: SharedPreferences
    private val client = OkHttpClient()
    private var refreshJob: Job? = null

    init {
        val masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
        prefs = EncryptedSharedPreferences.create(
            "molam_auth_prefs",
            masterKeyAlias,
            context,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    companion object {
        @Volatile
        private var instance: MolamAuth? = null

        fun init(context: Context, config: MolamAuthConfig): MolamAuth {
            return instance ?: synchronized(this) {
                instance ?: MolamAuth(context.applicationContext, config).also { instance = it }
            }
        }

        fun getInstance(): MolamAuth {
            return instance ?: throw IllegalStateException("MolamAuth not initialized. Call init() first.")
        }
    }

    // MARK: - Public API

    suspend fun login(request: LoginRequest): Result<AuthResponse> = withContext(Dispatchers.IO) {
        val deviceId = request.deviceId ?: getOrCreateDeviceId()

        val json = JSONObject().apply {
            put("identifier", request.identifier)
            put("password", request.password)
            put("device_id", deviceId)
            put("device_type", request.deviceType)
        }

        val result = makeRequest("/api/id/auth/login", json)

        result.fold(
            onSuccess = { response ->
                val authResponse = parseAuthResponse(response)
                setTokens(authResponse.accessToken, authResponse.refreshToken)

                if (config.autoRefresh) {
                    scheduleRefresh(authResponse.expiresIn)
                }

                Result.success(authResponse)
            },
            onFailure = { error ->
                Result.failure(error)
            }
        )
    }

    suspend fun refresh(): Result<AuthResponse> = withContext(Dispatchers.IO) {
        val refreshToken = getRefreshToken()
            ?: return@withContext Result.failure(Exception("No refresh token available"))

        val json = JSONObject().apply {
            put("refresh_token", refreshToken)
        }

        val result = makeRequest("/api/id/auth/refresh", json)

        result.fold(
            onSuccess = { response ->
                val authResponse = parseAuthResponse(response)
                setTokens(authResponse.accessToken, authResponse.refreshToken)

                if (config.autoRefresh) {
                    scheduleRefresh(authResponse.expiresIn)
                }

                Result.success(authResponse)
            },
            onFailure = { error ->
                clearTokens()
                Result.failure(error)
            }
        )
    }

    suspend fun logout(sessionId: String? = null): Result<Unit> = withContext(Dispatchers.IO) {
        val accessToken = getAccessToken()

        if (accessToken != null) {
            val json = JSONObject().apply {
                sessionId?.let { put("session_id", it) }
            }

            makeAuthenticatedRequest("/api/id/auth/logout", json, accessToken)
        }

        clearTokens()
        refreshJob?.cancel()
        refreshJob = null

        Result.success(Unit)
    }

    fun getAccessToken(): String? {
        return prefs.getString("access_token", null)
    }

    fun getRefreshToken(): String? {
        return prefs.getString("refresh_token", null)
    }

    fun isAuthenticated(): Boolean {
        return getAccessToken() != null
    }

    suspend fun authenticatedRequest(
        path: String,
        json: JSONObject,
        method: String = "POST"
    ): Result<String> = withContext(Dispatchers.IO) {
        val accessToken = getAccessToken()
            ?: return@withContext Result.failure(Exception("Not authenticated"))

        val result = makeAuthenticatedRequest(path, json, accessToken, method)

        // Auto-refresh on 401
        if (result.isFailure && result.exceptionOrNull()?.message?.contains("401") == true) {
            if (config.autoRefresh) {
                val refreshResult = refresh()
                if (refreshResult.isSuccess) {
                    val newToken = getAccessToken()!!
                    return@withContext makeAuthenticatedRequest(path, json, newToken, method)
                }
            }
        }

        result
    }

    // MARK: - Private Methods

    private fun setTokens(accessToken: String, refreshToken: String) {
        prefs.edit().apply {
            putString("access_token", accessToken)
            putString("refresh_token", refreshToken)
            apply()
        }
    }

    private fun clearTokens() {
        prefs.edit().apply {
            remove("access_token")
            remove("refresh_token")
            apply()
        }
    }

    private fun scheduleRefresh(expiresIn: Int) {
        refreshJob?.cancel()

        val refreshIn = maxOf((expiresIn - 60) * 1000L, 0L)

        refreshJob = CoroutineScope(Dispatchers.IO).launch {
            delay(refreshIn)
            refresh()
        }
    }

    private fun getOrCreateDeviceId(): String {
        val key = "device_id"
        var deviceId = prefs.getString(key, null)

        if (deviceId == null) {
            deviceId = "android-${UUID.randomUUID()}"
            prefs.edit().putString(key, deviceId).apply()
        }

        return deviceId
    }

    private suspend fun makeRequest(
        path: String,
        json: JSONObject,
        method: String = "POST"
    ): Result<String> = suspendCancellableCoroutine { continuation ->
        val url = "${config.baseUrl}$path"

        val requestBuilder = Request.Builder()
            .url(url)
            .addHeader("Content-Type", "application/json")

        config.apiKey?.let {
            requestBuilder.addHeader("X-API-Key", it)
        }

        val body = json.toString().toRequestBody("application/json".toMediaType())

        when (method) {
            "POST" -> requestBuilder.post(body)
            "PUT" -> requestBuilder.put(body)
            "PATCH" -> requestBuilder.patch(body)
            "DELETE" -> requestBuilder.delete(body)
            else -> requestBuilder.post(body)
        }

        client.newCall(requestBuilder.build()).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                continuation.resumeWith(Result.failure(e))
            }

            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    val responseBody = response.body?.string() ?: ""
                    continuation.resumeWith(Result.success(responseBody))
                } else {
                    continuation.resumeWith(
                        Result.failure(
                            Exception("HTTP ${response.code}: ${response.message}")
                        )
                    )
                }
            }
        })
    }

    private suspend fun makeAuthenticatedRequest(
        path: String,
        json: JSONObject,
        accessToken: String,
        method: String = "POST"
    ): Result<String> = suspendCancellableCoroutine { continuation ->
        val url = "${config.baseUrl}$path"

        val requestBuilder = Request.Builder()
            .url(url)
            .addHeader("Content-Type", "application/json")
            .addHeader("Authorization", "Bearer $accessToken")

        val body = json.toString().toRequestBody("application/json".toMediaType())

        when (method) {
            "POST" -> requestBuilder.post(body)
            "PUT" -> requestBuilder.put(body)
            "PATCH" -> requestBuilder.patch(body)
            "DELETE" -> requestBuilder.delete(body)
            else -> requestBuilder.post(body)
        }

        client.newCall(requestBuilder.build()).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                continuation.resumeWith(Result.failure(e))
            }

            override fun onResponse(call: Call, response: Response) {
                if (response.isSuccessful) {
                    val responseBody = response.body?.string() ?: ""
                    continuation.resumeWith(Result.success(responseBody))
                } else {
                    continuation.resumeWith(
                        Result.failure(
                            Exception("HTTP ${response.code}: ${response.message}")
                        )
                    )
                }
            }
        })
    }

    private fun parseAuthResponse(jsonString: String): AuthResponse {
        val json = JSONObject(jsonString)
        val userJson = json.getJSONObject("user")

        return AuthResponse(
            accessToken = json.getString("access_token"),
            refreshToken = json.getString("refresh_token"),
            expiresIn = json.getInt("expires_in"),
            tokenType = json.getString("token_type"),
            user = AuthResponse.User(
                id = userJson.getString("id"),
                email = userJson.optString("email").takeIf { it.isNotEmpty() },
                phoneE164 = userJson.optString("phone_e164").takeIf { it.isNotEmpty() }
            )
        )
    }
}

File: brique-24-sdk-auth/sdk/js/molam-auth.ts
/**
 * Molam Auth SDK - JavaScript/TypeScript
 * Universal authentication SDK for Molam ID
 *
 * @version 1.0.0
 * @author Molam
 */

export interface MolamAuthConfig {
  baseUrl: string;
  apiKey?: string;
  storage?: 'localStorage' | 'sessionStorage' | 'memory';
  autoRefresh?: boolean;
  onTokenRefresh?: (token: string) => void;
  onSessionExpired?: () => void;
}

export interface LoginRequest {
  identifier: string;
  password: string;
  device_id?: string;
  device_type?: 'web' | 'mobile' | 'desktop' | 'api';
}

export interface AuthResponse {
  access_token: string;
  refresh_token: string;
  expires_in: number;
  token_type: string;
  user: {
    id: string;
    email?: string;
    phone_e164?: string;
  };
}

export class MolamAuth {
  private config: MolamAuthConfig;
  private accessToken: string | null = null;
  private refreshToken: string | null = null;
  private refreshTimer: any = null;

  constructor(config: MolamAuthConfig) {
    this.config = {
      storage: 'localStorage',
      autoRefresh: true,
      ...config,
    };
    this.loadTokens();
  }

  /**
   * Login with email/phone and password
   */
  async login(request: LoginRequest): Promise<AuthResponse> {
    const device_id = request.device_id || this.generateDeviceId();
    const device_type = request.device_type || 'web';

    const response = await this.request('/api/id/auth/login', {
      method: 'POST',
      body: JSON.stringify({
        identifier: request.identifier,
        password: request.password,
        device_id,
        device_type,
      }),
    });

    const data: AuthResponse = await response.json();
    this.setTokens(data.access_token, data.refresh_token);

    if (this.config.autoRefresh) {
      this.scheduleRefresh(data.expires_in);
    }

    return data;
  }

  /**
   * Refresh access token
   */
  async refresh(): Promise<AuthResponse> {
    if (!this.refreshToken) {
      throw new Error('No refresh token available');
    }

    const response = await this.request('/api/id/auth/refresh', {
      method: 'POST',
      body: JSON.stringify({
        refresh_token: this.refreshToken,
      }),
    });

    if (!response.ok) {
      this.clearTokens();
      if (this.config.onSessionExpired) {
        this.config.onSessionExpired();
      }
      throw new Error('Failed to refresh token');
    }

    const data: AuthResponse = await response.json();
    this.setTokens(data.access_token, data.refresh_token);

    if (this.config.autoRefresh) {
      this.scheduleRefresh(data.expires_in);
    }

    return data;
  }

  /**
   * Logout
   */
  async logout(sessionId?: string): Promise<void> {
    try {
      await this.request('/api/id/auth/logout', {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${this.accessToken}`,
        },
        body: JSON.stringify({ session_id: sessionId }),
      });
    } catch (err) {
      console.error('Logout error:', err);
    } finally {
      this.clearTokens();
      if (this.refreshTimer) {
        clearTimeout(this.refreshTimer);
        this.refreshTimer = null;
      }
    }
  }

  /**
   * Get current access token
   */
  getAccessToken(): string | null {
    return this.accessToken;
  }

  /**
   * Check if user is authenticated
   */
  isAuthenticated(): boolean {
    return !!this.accessToken;
  }

  /**
   * Make authenticated request
   */
  async authenticatedRequest(url: string, options: RequestInit = {}): Promise<Response> {
    if (!this.accessToken) {
      throw new Error('Not authenticated');
    }

    const headers = {
      ...options.headers,
      Authorization: `Bearer ${this.accessToken}`,
    };

    const response = await this.request(url, { ...options, headers });

    // Auto-refresh on 401
    if (response.status === 401 && this.config.autoRefresh) {
      await this.refresh();
      headers.Authorization = `Bearer ${this.accessToken}`;
      return this.request(url, { ...options, headers });
    }

    return response;
  }

  // Private methods

  private async request(path: string, options: RequestInit = {}): Promise<Response> {
    const url = `${this.config.baseUrl}${path}`;
    const headers = {
      'Content-Type': 'application/json',
      ...(this.config.apiKey && { 'X-API-Key': this.config.apiKey }),
      ...options.headers,
    };

    return fetch(url, { ...options, headers });
  }

  private setTokens(accessToken: string, refreshToken: string): void {
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.storeTokens();

    if (this.config.onTokenRefresh) {
      this.config.onTokenRefresh(accessToken);
    }
  }

  private clearTokens(): void {
    this.accessToken = null;
    this.refreshToken = null;
    this.removeTokens();
  }

  private storeTokens(): void {
    if (this.config.storage === 'memory') {
      return;
    }

    const storage = this.config.storage === 'sessionStorage' ? sessionStorage : localStorage;
    if (this.accessToken) storage.setItem('molam_access_token', this.accessToken);
    if (this.refreshToken) storage.setItem('molam_refresh_token', this.refreshToken);
  }

  private loadTokens(): void {
    if (this.config.storage === 'memory') {
      return;
    }

    const storage = this.config.storage === 'sessionStorage' ? sessionStorage : localStorage;
    this.accessToken = storage.getItem('molam_access_token');
    this.refreshToken = storage.getItem('molam_refresh_token');
  }

  private removeTokens(): void {
    if (this.config.storage === 'memory') {
      return;
    }

    const storage = this.config.storage === 'sessionStorage' ? sessionStorage : localStorage;
    storage.removeItem('molam_access_token');
    storage.removeItem('molam_refresh_token');
  }

  private scheduleRefresh(expiresIn: number): void {
    if (this.refreshTimer) {
      clearTimeout(this.refreshTimer);
    }

    // Refresh 1 minute before expiry
    const refreshIn = Math.max((expiresIn - 60) * 1000, 0);
    this.refreshTimer = setTimeout(() => {
      this.refresh().catch(console.error);
    }, refreshIn);
  }

  private generateDeviceId(): string {
    let deviceId = localStorage.getItem('molam_device_id');
    if (!deviceId) {
      deviceId = `web-${Math.random().toString(36).substring(2, 15)}`;
      localStorage.setItem('molam_device_id', deviceId);
    }
    return deviceId;
  }
}

// Export singleton instance
let defaultInstance: MolamAuth | null = null;

export function initMolamAuth(config: MolamAuthConfig): MolamAuth {
  defaultInstance = new MolamAuth(config);
  return defaultInstance;
}

export function getMolamAuth(): MolamAuth {
  if (!defaultInstance) {
    throw new Error('MolamAuth not initialized. Call initMolamAuth() first.');
  }
  return defaultInstance;
}

export default MolamAuth;
